File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {

        int n = heights.length;

        // Next Smallest Left
        int[] NSL = new int[n];
        // Next Smallest Right
        int[] NSR = new int[n];

        for(int i=0 ; i<n ; i++){
            NSL[i] = -1;
            NSR[i] = n;
        }

        Stack<Integer> stack = new Stack<>();

        for(int i = 0 ; i < n ; i++){
            while(!stack.empty() && heights[stack.peek()] > heights[i]){
                NSR[stack.peek()] = i;
                stack.pop();
            }
            stack.push(i);
        }

        stack.clear();

        for(int i = 0 ; i < n ; i++){
            while(!stack.empty() && heights[stack.peek()] > heights[i]){
                stack.pop();
            }
            if(!stack.empty()){
                NSL[i] = stack.peek();
            }

            stack.push(i);
        }

        int max_area = 0;

        for(int i=0 ; i<n ; i++){
            int curr_height = heights[i];
            int width = NSR[i] - NSL[i] - 1;
            int curr_area = curr_height * width;

            if(curr_area>max_area){
                max_area = curr_area;
            }

        }


        return max_area;
        
    }
}

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        
        int n = heights.size();
        int max_area = 0;
        stack<int> stack;

        for(int i=0 ; i<n ; i++){
            
            while(!stack.empty() && heights[stack.top()] > heights[i] ){
                int curr_height = heights[stack.top()];
                int NSR = i;
                int NSL;
                stack.pop();
                if(stack.empty()){
                    NSL = -1;
                }
                else{
                    NSL = stack.top();
                }

                int curr_area = curr_height * (NSR - NSL - 1);
                if(curr_area > max_area){
                    max_area = curr_area;
                }

            }
            stack.push(i);

        }

        while(!stack.empty()){
            int NSR = n;
            int NSL;

            int curr_height = heights[stack.top()];

            stack.pop();
            if(stack.empty()){
                NSL = -1;
            }
            else{
                NSL = stack.top();
            }

            int curr_area = curr_height * (NSR - NSL - 1);
            
            if(curr_area > max_area){
                max_area = curr_area;
            }
        }

        return max_area;

    }
};